src/pmdas/mssql/pmdamssql.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2019-2021 Red Hat.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH PMDAMSSQL 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmdamssql\f1 \- Microsoft SQL Server database PMDA
.SH DESCRIPTION
\fBpmdamssql\fP is a Performance Co-Pilot (PCP) Performance Metrics
Domain Agent (PMDA) which extracts live performance data from a running
Microsoft SQL Server database server.
.SH CONFIGURATION
Firstly, \f3pmdamssql\f1 requires installation of these support packages:
.TP 5
.B mssql-server
The primary Microsoft SQL Server database package.
.TP
\fBmsodbcsql18\fR, \fBmsodbcsql17\fR or \fBmsodbcsql\fR
Microsoft SQL Server ODBC bindings.
.TP
.B pyodbc
General Python ODBC module with Microsoft SQL Server support enabled.
.PP
Next, \fBpmdamssql\fP reads an optional ini-style configuration file:
.IP
.PD 0
.IP
.I \f(CR$PCP_PMDAS_DIR\fP/mssql/mssql.conf
.PD
.PP
This file currently allows for two sections, one for configuring the
library that connects to SQL server, the other for authenticating with
SQL server.
.PP
.B [connection]
section specifies values for the following settings
(their default values are shown in parenthesis):
.TP 15
.B driver \fR(\fP\fI{ODBC Driver 17 for SQL Server}\fP\fR)\fP
ODBC driver string.
.TP
.B server \fR(\fP\fItcp:localhost\fP\fR)\fP
SQL Server socket connection string.
.TP
.B timeout \fR(\fP\fI2\fP\fR)\fP
Connection timeout setting, in seconds.
.PP
.B [authentication]
section specifies values for the two supported
modes of authentication with SQL Server.
.P
The default is to use SQL Server authentication,
which requires a username and password to be
provided.
These are the following configuration settings
(their default values are shown in parenthesis):
.TP 15
.B username \fR(\fP\fIpcp\fP\fR)\fP
Username to connect to the database.
.TP
.B password \fR(empty)\fP
Password to connect to the database.
.PP
Note that if a SQL Server Assessments API configuration file is
found at \fI/var/opt/mssql/secrets/assessment\fP, then
.B pmdamssql
reads the username and password from that file preferentially.
The format is simply a two line text file, the first containing
the username and the second the password.
.PP
The second option is Windows authentication mode,
where logins are created in SQL Server that are not
based on operating system user accounts.
.TP 15
.B trusted \fR(\fP\fIfalse\fP\fR)\fP
Set to
.I true
to use Windows authentication when connecting to the
database.
.PP
This is a Kerberos-style connection and credentials
must be available before the PMDA is started (see
.BR kinit (1)
for further details).
.PD
.SH INSTALLATION
To install, the following must be done as root:
.sp 1
.RS +4
.ft B
.nf
# cd $PCP_PMDAS_DIR/mssql
# ./Install
.fi
.ft P
.RE
.sp 1
To uninstall, the following must be done as root:
.sp 1
.RS +4
.ft B
.nf
# cd $PCP_PMDAS_DIR/mssql
# ./Remove
.fi
.ft P
.RE
.sp 1
\fBpmdamssql\fP is launched by \fBpmcd\fP(1) and should never be
executed directly.
The \fBInstall\fP and \fBRemove\fP scripts notify \fBpmcd\fP(1) when
the agent is installed or removed.
.PP
\fBpmdamssql\fR will automatically attempt to reconnect to the SQL Server
if its connection is lost - e.g. if the SQL Server service is restarted.
.SH FILES
.TP 5
.I \f(CR$PCP_PMDAS_DIR\fP/mssql/mssql.conf
configuration file for the \fBpmdamssql\fR agent
.TP
.I \f(CR$PCP_PMDAS_DIR\fP/mssql/Install
installation script for the \fBpmdamssql\fR agent
.TP
.I \f(CR$PCP_PMDAS_DIR\fP/mssql/Remove
undo installation script for the \fBpmdamssql\fR agent
.TP
.I \f(CR$PCP_LOG_DIR\fP/pmcd/mssql.log
default log file for messages from the \fBpmdamssql\fR agent
.PP
Note that the usual/default value for \fB$PCP_PMDAS_DIR\fP is
.B /var/lib/pcp/pmdas
and the default for \fB$PCP_LOG_DIR\fP is
.B /var/log/pcp
but these settings are platform dependent.
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fR are used to parameterize
the file and directory names used by PCP.
On each installation, the
file \fI/etc/pcp.conf\fR contains the local values for these variables.
The \fB$PCP_CONF\fR variable may be used to specify an alternative
configuration file, as described in \fBpcp.conf\fR(5).
.SH SEE ALSO
.BR PCPIntro (1),
.BR kinit (1)
and
.BR pmcd (1).

.\" control lines for scripts/man-spell
.\" +ok+ pmdamssql msodbcsql pyodbc kinit mssql ODBC tcp


CHANGELOG
pcp-6.3.6 (17 March 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.4 (13 March 2025)
    - libpcp: fix endian issue in big-endian __pmTimestamp writing
    - pcp-xsos: fine-tune error handling on bad command line options
    - pcp-xsos: single-line pmPrintValue/pminfo values, escape shell chars
    - pcp-atop: correct BPF per-process network metric names
    - pmcd, pmdaproc: allow non-local username connection attributes
    - pmlogger, libpcp: fix handling of dynamic instance domain labels
    - pmrep: use BPF instead of BCC for proc networking metrics
    - pmlogconf: use libbpf over BCC for networking metrics
    - pmlogextract: fix logic around instances labels, allowing set updates
    - pmdalinux: mem.numa.hugepages metrics (per node, per hugepage size)
    - pmdabpftrace: fix for some python3 versions async IO create_task
    - pmdanvidia: if present use dso from the libnvidia-ml runtime package
    - pmdadm, pmieconf: default to enabling the dmthin pmie rules
    - configure.ac: fix autoreconf, add AC_DEFINE_UNQUOTED descriptions
    - rpm builds: enable bpf, bpftrace, bcc on riscv64 builds
    - rpm builds: drop user creation scriptlets for rpm builds
    - deb build: force debian/rules to be recreated each time
    - containers: updates to the archive-analysis container
    - selinux: several policy updates for new AVC denials
    - docs: improve SSL/TLS diagnostic in pcp-verify script

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.2 (6 November 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdabpf: avoid a too-low rlimit setting within pmdabpf
	pmdaproc: add an access control mechanism for remote sampling
	pmdalinux: ensure divide-by-zero cannot happen with tmpfs.full
	pmdalinux: additional MPTcpExt netstat metrics
	pmdastatsd: keep correct track of pmns mapped metric count
	pmdaopenmetrics: run as pcp user rather than root by default
	pmdaopenmetrics: ensure correct cache file permissions used
    - Server-side utilities and log management scripts:
	libpcp, pmlogger: add support for zstd(1) compression of archives
	pmlogbasename: new, maps filename to PCP archive basename
	pmlogcompress: new, archive compression/decompression utility
	pmlogcheck: performance improvements
	pmlogcheck:  allow multiple archive options on command line
	pmlogextract: extend -v option with the full pmlogger syntax
	pmlogextract: fix for reporting file size on 32-bit platforms
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix 64-bit big endian v3 archive timestamp decoding
	libpcp: change in archive value interpolation for non-counter metrics
	python api: allow timeval and timespec init with float/datetime
    - Client tools and utilities:
	pminfo: minor error path reworking
	pcp-buddyinfo: addition of a --no-interpol option
	pcp-ps: added option to view full task name with its argument
    - Build, infrastructure and packaging updates:
	build: deal with botched deb package names from Debian
	build: ensure pmie config.default created with correct ownership
	systemd: move pcp-reboot-init rpm spec logic from zeroconf to pcp
	selinux: ensure farm pmlogger/pmie processes labelled correctly
    - Documentation:
	docs: refine pmdaproc -A description details
	docs: make pmstat(1) match the code for -t defaults

pcp-6.3.1 (17 September 2024)
    - Client tools and utilities:
	pcp2arrow: fix extraction of values for metrics with indoms
	pcp-htop: fix platform setting enabling screen tabs
	pcp-htop: update to latest upstream bug fixes
	pcp-meminfo: added timestamp string for consistency
    - PMDA additions, enhancements and bug fixes:
	pmdaopenmetrics: add vLLM metrics support by default
	pmdahacluster: handle pacemaker 2.1.6 crm_mon output changes
	pmdahacluster: handle new style Pacemaker role descriptions
	pmdalinux: MPJoinSynBackupRx, MPJoinSynAckBackupRx netstat metrics
	pmdalinux: sysfs fallback from hinv.cpu.clock metric
	pmdaproc: lazy initialisation for accounting metrics
	pmdapmcd: expose pmcd limits as new metrics
	pmdammv: use isnan instead of open-coding NaN checks
    - Server-side utilities and log management scripts:
	pmcd: put an explicit cap on max number of fetched PMIDs
	pmcd: put an explicit cap on max number of contexts per client
	pmcd: cleanup caching of local hostname string
	pmpost: guard against possible symlink attack
	pmproxy: transition to dual Redis/Valkey support
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp_web: fix a race in webgroup context teardown on timeout
	libpcp_web: transition to dual Redis/Valkey support
	libpcp: auto-load the proc PMDA in local context mode
	libpcp: fix buffer overflow in __pmServerAddInterface
	libpcp: refactor the attr PDU content checking
	libpcp: desc PDU consistency check and error message rework
	libpcp: text PDU consistency check and error message rework
	libpcp: creds PDU consistency check and error message rework
	libpcp: label PDU consistency check and error message rework
	libpcp: instance PDU consistency check and error message rework
	libpcp: rework some fetch PDU integrity checks and error messages
	libpcp: reworked __pmDecodeIDList integrity checks and warnings
	libpcp: ensure no fd leakage to child from __pmProcessPipe
	libpcp: ensure container names contain limited characters
	libpcp: fix buffer overrun test in __pmDecodeValueSet
	libpcp: hardening of the results PDU handling
	libpcp: hardening of the creds PDU handling
	libpcp: add new __pmCtlDebug() function
	python API: fix MMV metric text, indom and semantics handling
	python API: add log import timestamp convenience mechanisms
    - Build, infrastructure and packaging updates:
	build: tweak Makepkgs as dpkg-parsechangelog(1) is a bit picky
	build: drop gfs2 PMDA from build if distro vendor opts-out

src/pmdas/mssql/pmdamssql.pl
#
# Copyright (c) 2011 Aconex.  All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#

use strict;
use warnings;
use PCP::PMDA;
use DBI;

my $server = 'localhost';
my $database = 'PCP';
my $username = 'dbmonitor';
my $password = 'dbmonitor';

# Configuration files for overriding the above settings
for my $file (	'/etc/pcpdbi.conf',	# system defaults (lowest priority)
		pmda_config('PCP_PMDAS_DIR') . '/mssql/mssql.conf',
		'./mssql.conf' ) {	# current directory (high priority)
    eval `cat $file` unless ! -f $file;
}

use vars qw( $pmda $dbh );
use vars qw( $sth_os_memory_clerks  );
use vars qw( $sth_virtual_file_stats @virtual_file_stats );
use vars qw( $sth_total_running_user_processes @total_running_user_processes );
use vars qw( $sth_os_memory_clerks @os_memory_clerks );
use vars qw( $sth_os_workers_waiting_cpu @os_workers_waiting_cpu );
my $database_indom = 0;
my @database_instances;

sub mssql_connection_setup
{
    #$pmda->log("mssql_connection_setup\n");

    if (!defined($dbh)) {
    	$dbh = DBI->connect("DBI:Sybase:server=$server", $username, $password);
    	if (defined($dbh)) {
	        $pmda->log("MSSQL connection established\n");
	        $sth_virtual_file_stats = $dbh->prepare(
        	    "select db_name(database_id), cast(num_of_reads as numeric), cast(num_of_bytes_read as numeric)," .
        	    " cast(io_stall_read_ms as numeric), cast(num_of_writes as numeric)," .
        	    " cast(num_of_bytes_written as numeric), cast(io_stall_write_ms as numeric)," .
        	    " cast(size_on_disk_bytes as numeric) " .
        	    "from sys.dm_io_virtual_file_stats(DB_ID(''),1)");
	        $sth_os_memory_clerks = $dbh->prepare(
	             "SELECT SUM(multi_pages_kb + virtual_memory_committed_kb + shared_memory_committed_kb + awe_allocated_kb)" .
	             " from sys.dm_os_memory_clerks WHERE type IN " .
	             "('MEMORYCLERK_SQLBUFFERPOOL', 'MEMORYCLERK_SQLQUERYCOMPILE'," .
	             " 'MEMORYCLERK_SQLQUERYEXEC', 'MEMORYCLERK_SQLQUERYPLAN')" .
	             " group by type order by type");
	        $sth_total_running_user_processes = $dbh->prepare(
	             "SELECT count(*) FROM sys.dm_exec_requests " .
	             "WHERE session_id >= 51 AND status = 'running'");
            $sth_os_workers_waiting_cpu = $dbh->prepare(
	             "SELECT ISNULL(COUNT(*),0) FROM sys.dm_os_workers AS workers " .
	             "INNER JOIN sys.dm_os_schedulers AS schedulers " .
	             "ON workers.scheduler_address = schedulers.scheduler_address " .
	             "WHERE workers.state = 'RUNNABLE' AND schedulers.scheduler_id < 255"); 
	    }
    }
}

sub mssql_os_memory_clerks_refresh
{
    #$pmda->log("mssql_os_memory_clerks_refresh\n");

    @os_memory_clerks = ();	# clear any previous contents
    if (defined($dbh)) {
    	$sth_os_memory_clerks->execute();
        my $result = $sth_os_memory_clerks->fetchall_arrayref();
	    for my $i (0 .. $#{$result}) {	    
	        $os_memory_clerks[$i] = $result->[$i][0];
	    }
	}
}

sub mssql_virtual_file_stats_refresh
{
    #$pmda->log("mssql_virtual_file_stats_refresh\n");

    @virtual_file_stats = ();	# clear any previous contents
    @database_instances = ();

    if (defined($dbh)) {
    	$sth_virtual_file_stats->execute();
	    my $result = $sth_virtual_file_stats->fetchall_arrayref();

	    for my $i (0 .. $#{$result}) {
	        $database_instances[($i*2)] = $i;
	        $database_instances[($i*2)+1] = "$result->[$i][0]";
	        $virtual_file_stats[$i] = $result->[$i];
	    }

	    $pmda->replace_indom( $database_indom, \@database_instances );
    }
}

sub mssql_total_running_user_processes
{
    #$pmda->log("mssql_total_running_user_processes\n");

    @total_running_user_processes = ();	# clear any previous contents
    if (defined($dbh)) {
    	$sth_total_running_user_processes->execute();
	my $result = $sth_total_running_user_processes->fetchall_arrayref();
	@total_running_user_processes = ( $result->[0][0] );
    }
}

sub mssql_os_workers_waiting_cpu_refresh
{
    #$pmda->log("mssql_os_workers_refresh\n");

    @os_workers_waiting_cpu = ();	# clear any previous contents
    if (defined($dbh)) {
    	$sth_os_workers_waiting_cpu->execute();
        my $result = $sth_os_workers_waiting_cpu->fetchall_arrayref();
        @os_workers_waiting_cpu = ( $result->[0][0] );
	}
}

sub mssql_refresh
{
    my ($cluster) = @_;

    #$pmda->log("mssql_refresh $cluster\n");

    if ($cluster == 0)		{ mssql_virtual_file_stats_refresh; }
    elsif ($cluster == 1)	{ mssql_os_memory_clerks_refresh; }
    elsif ($cluster == 2)	{ mssql_total_running_user_processes; }
    elsif ($cluster == 3)	{ mssql_os_workers_waiting_cpu_refresh; }
}

sub mssql_fetch_callback
{
    my ($cluster, $item, $inst) = @_;
    my ($value, @vfstats);

    #$pmda->log("mssql_fetch_callback $cluster:$item ($inst)\n");

    if ($cluster == 0) {
        if ($item > 6)              { return (PM_ERR_PMID, 0); }
    	if ($inst < 0)		        { return (PM_ERR_INST, 0); }
    	if ($inst > @database_instances)	{ return (PM_ERR_INST, 0); }
       	$value = $virtual_file_stats[$inst];
       	if (!defined($value))	    { return (PM_ERR_INST, 0); }
       	@vfstats = @$value;
        if (!defined($vfstats[$item+1])) { return (PM_ERR_AGAIN, 0); }
        return ($vfstats[$item+1], 1);
    }
    if ($inst != PM_IN_NULL)		{ return (PM_ERR_INST, 0); }
    if ($cluster == 1) {
        if ($item > 3)              { return (PM_ERR_PMID, 0); }
        if (!defined($os_memory_clerks[$item])) { return (PM_ERR_AGAIN, 0); }
        return ($os_memory_clerks[$item], 1);
    }
    if ($cluster == 2) {
        if ($item > 0)              { return (PM_ERR_PMID, 0); }
        if (!defined($total_running_user_processes[$item])) { return (PM_ERR_AGAIN, 0); }
        return ($total_running_user_processes[$item], 1);
    }
    if ($cluster == 3) {
        if ($item > 0)              { return (PM_ERR_PMID, 0); }
        if (!defined($os_workers_waiting_cpu[$item])) { return (PM_ERR_AGAIN, 0); }
        return ($os_workers_waiting_cpu[$item], 1);
    }
    return (PM_ERR_PMID, 0);
}

$pmda = PCP::PMDA->new('mssql', 109);

$pmda->add_metric(pmda_pmid(0,0), PM_TYPE_U64, $database_indom,
		  PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
		  'mssql.virtual_file.read', 'Number of bytes reads issued on data file', '');
$pmda->add_metric(pmda_pmid(0,1), PM_TYPE_U64, $database_indom,
		  PM_SEM_COUNTER, pmda_units(1,0,0,PM_SPACE_BYTE,0,0),
		  'mssql.virtual_file.read_bytes', 'Total number of bytes read on the data file', '');
$pmda->add_metric(pmda_pmid(0,2), PM_TYPE_U64, $database_indom,
		  PM_SEM_COUNTER, pmda_units(0,1,0,0,PM_TIME_MSEC,0),
		  'mssql.virtual_file.read_io_stall_time', 'Total time in ms that the users waited for reads issued on the file', '');
$pmda->add_metric(pmda_pmid(0,3), PM_TYPE_U64, $database_indom,
		  PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
		  'mssql.virtual_file.write', 'Number of writes made on the data file', '');
$pmda->add_metric(pmda_pmid(0,4), PM_TYPE_U64, $database_indom,
		  PM_SEM_COUNTER, pmda_units(1,0,0,PM_SPACE_BYTE,0,0),
		  'mssql.virtual_file.write_bytes', 'Total number of bytes written to the data file', '');
$pmda->add_metric(pmda_pmid(0,5), PM_TYPE_U64, $database_indom,
		  PM_SEM_COUNTER, pmda_units(0,1,0,0,PM_TIME_MSEC,0),
		  'mssql.virtual_file.write_io_stall_time', 'Total time in ms that users waited for writes to be completed o the file', '');
$pmda->add_metric(pmda_pmid(0,6), PM_TYPE_U64, $database_indom,
		  PM_SEM_INSTANT, pmda_units(1,0,0,PM_SPACE_BYTE,0,0),
		  'mssql.virtual_file.size', 'Number of bytes used on the disk from the data file', '');

$pmda->add_metric(pmda_pmid(1,0), PM_TYPE_U32, PM_INDOM_NULL,
		  PM_SEM_INSTANT, pmda_units(1,0,0,PM_SPACE_KBYTE,0,0),
		  'mssql.os_memory_clerks.bufferpool', '', '');
$pmda->add_metric(pmda_pmid(1,1), PM_TYPE_U32, PM_INDOM_NULL,
		  PM_SEM_INSTANT, pmda_units(1,0,0,PM_SPACE_KBYTE,0,0),
		  'mssql.os_memory_clerks.querycompile', '', '');
$pmda->add_metric(pmda_pmid(1,2), PM_TYPE_U32, PM_INDOM_NULL,
		  PM_SEM_INSTANT, pmda_units(1,0,0,PM_SPACE_KBYTE,0,0),
		  'mssql.os_memory_clerks.queryexec', '', '');
$pmda->add_metric(pmda_pmid(1,3), PM_TYPE_U32, PM_INDOM_NULL,
		  PM_SEM_INSTANT, pmda_units(1,0,0,PM_SPACE_KBYTE,0,0),
		  'mssql.os_memory_clerks.queryplan', '', '');

$pmda->add_metric(pmda_pmid(2,0), PM_TYPE_U32, PM_INDOM_NULL,
		  PM_SEM_INSTANT, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
		  'mssql.running_user_process.total', 'Total number of running user process belonging to aconexsq', '');

$pmda->add_metric(pmda_pmid(3,0), PM_TYPE_U32, PM_INDOM_NULL,
		  PM_SEM_INSTANT, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
		  'mssql.os_workers_waiting_cpu.count', 'Total number of queries waiting for cpu', '');

$pmda->add_indom($database_indom, \@database_instances,
    	 'Instance domain exporting each MSSQL database', '');

$pmda->set_fetch_callback(\&mssql_fetch_callback);
$pmda->set_fetch(\&mssql_connection_setup);
$pmda->set_refresh(\&mssql_refresh);
$pmda->run;

=pod

=head1 NAME

pmdamssql - Microsoft SQL database PMDA

=head1 DESCRIPTION

B<pmdamssql> is a Performance Co-Pilot PMDA which extracts
live performance data from a running SQL Server database.
These metrics are typically sourced from Dynamic Management
Views (DMVs), augmenting the SQL server metrics exported by
the Windows PMDA.

=head1 INSTALLATION

B<pmdamssql> uses a configuration file from (in this order):

=over

=item * /etc/pcpdbi.conf

=item * $PCP_PMDAS_DIR/mssql/mssql.conf

=back

This file can contain overridden values (Perl code) for the settings
listed at the start of pmdamssql.pl, namely:

=over

=item * database name (see DBI(3) for details)

=item * database user name

=item * database pass word

=back

Once this is setup, you can access the names and values for the
mysql performance metrics by doing the following as root:

	# cd $PCP_PMDAS_DIR/mssql
	# ./Install

If you want to undo the installation, do the following as root:

	# cd $PCP_PMDAS_DIR/mssql
	# ./Remove

B<pmdamssql> is launched by pmcd(1) and should never be executed
directly.  The Install and Remove scripts notify pmcd(1) when
the agent is installed or removed.

=head1 FILES

=over

=item /etc/pcpdbi.conf

configuration file for all PCP database monitors

=item $PCP_PMDAS_DIR/mssql/mssql.conf

configuration file for B<pmdamssql>

=item $PCP_PMDAS_DIR/mssql/Install

installation script for the B<pmdamssql> agent

=item $PCP_PMDAS_DIR/mssql/Remove

undo installation script for the B<pmdamssql> agent

=item $PCP_LOG_DIR/pmcd/mssql.log

default log file for error messages from B<pmdamssql>

=back

=head1 SEE ALSO

pmcd(1), pmdadbping.pl(1) and DBI(3).
